### PR TITLE
[ENHANCEMENT] added temp table option for snowflake

### PR DESCRIPTION
**Pull request by talagluck**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Added option to use temp tables in snowflake in addition to the transient table option
- If the `"snowflake_transient_table"` key is present in batch_kwargs, we will use a transient table (which persists), otherwise we will default to a temp table
- To do this, we execute from a connection rather than an engine (as we do with sqlite and mssql)

Previous Design Review notes:
- After discussion with Eugene, we decided to continue to provide an option to use  transient tables so that this would not introduce a breaking change.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 1]
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
                 table_name=table_name, custom_sql=custom_sql
             )
         elif self.sql_engine_dialect.name.lower() == "snowflake":
-            logger.info("Creating transient table %s" % table_name)
+            if not self.generated_table_name:
+                table_type = "TRANSIENT"
+            else:
+                table_type = "TEMPORARY"
+            logger.info("Creating temporary table %s" % table_name)
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py:1194]: Just a suggestion to convert to f-strings when changing/adding an interpolated string (https://www.python.org/dev/peps/pep-0498/)
```suggestion
            logger.info(f"Creating temporary table {table_name}")
```

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 2]
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
...
             if schema_name is not None:
                 table_name = schema_name + "." + table_name
-            stmt = "CREATE OR REPLACE TRANSIENT TABLE {table_name} AS {custom_sql}".format(
-                table_name=table_name, custom_sql=custom_sql
+            stmt = "CREATE OR REPLACE {table_type} TABLE {table_name} AS {custom_sql}".format(
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py:1197]:
```suggestion
            stmt = f"CREATE OR REPLACE {table_type} TABLE {table_name} AS {custom_sql}"
```

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 3]
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
...
+                table_type=table_type, table_name=table_name, custom_sql=custom_sql
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py:1198]:
```suggestion
                
```

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 4]
```diff
@@ -1222,6 +1222,10 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
             stmt = 'CREATE TEMPORARY TABLE "{table_name}" AS {custom_sql}'.format(
                 table_name=table_name, custom_sql=custom_sql
             )
+
+        # close_with_result = True
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py]: You might want to remove these commented out lines

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 5]
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
                 table_name=table_name, custom_sql=custom_sql
             )
         elif self.sql_engine_dialect.name.lower() == "snowflake":
-            logger.info("Creating transient table %s" % table_name)
+            if not self.generated_table_name:
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Totally up to you on style, but you could do these 4 lines with a ternary assignment:
`table_type = "TEMPORARY" if self.generated_table_name else "TRANSIENT"`

**Comment by anthonyburdi** [comment]
@talagluck Just a few suggestions! Take em or leave em.
@eugmandel Just to be clear I only did a code review, I did not consider the design.

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 6]
```diff
@@ -461,9 +461,9 @@ def __init__(
             # mssql expects all temporary table names to have a prefix '#'
             if engine.dialect.name.lower() == "mssql":
                 table_name = f"#{table_name}"
-            generated_table_name = table_name
+            self.generated_table_name = table_name
```
Comment (anthonyburdi) [inline great_expectations/dataset/sqlalchemy_dataset.py:466]: 👍

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 4 above)
Comment (talagluck) [reply great_expectations/dataset/sqlalchemy_dataset.py]: Great catch - thanks!

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 5 above)
Comment (talagluck) [reply great_expectations/dataset/sqlalchemy_dataset.py]: Great suggestion - thanks!

**Review by eugmandel — COMMENTED** [review]
@talagluck The code changes look good to me.

I don't see any changes in the CLI. Currently, it prompts for the transient table name. What do you think we should do? My thinking is to not require it in the CLI and have users configure it in the code because it will become a minority option.

Also, the how to guide for adding a Snowflake Datasource should reflect this change.

**Comment by talagluck** [comment]
Thanks, @eugmandel. Great point about the CLI - I updated the CLI to remove this prompt entirely.

I also updated the guide for adding a Snowflake Datasource.

**Review by eugmandel — APPROVED** [review]
@talagluck Thank you for getting the CLI and the guide changes in! I see Anthony has some code edit suggesting. Approving - your judgement which suggestions to accept before merging.

**Pull request merged**